File: python/739-daily-temperatures.py
# Time -> O(n), Space -> O(n)

class Solution:
    def dailyTemperatures(self, temperatures: List[int]) -> List[int]:
        stack = []
        result = [0] * len(temperatures)

        for i, value in enumerate(temperatures):
            while (len(stack) and value > temperatures[stack[-1]]):
                top = stack.pop()
                result[top] = i - top
            stack.append(i)

        return result

File: cpp/739-daily-temperatures.cpp
// Time -> O(n), Space -> O(n)

#include<vector>
#include <stack>
using namespace std;

class Solution {
public:
    vector<int> dailyTemperatures(vector<int>& temperatures) {
        stack<int> s;
        vector<int> result(temperatures.size(), 0);

        for (int i = 0; i < temperatures.size(); i++) {
            while(!s.empty() && temperatures[s.top()] < temperatures[i]) {
                result[s.top()] = i - s.top();
                s.pop();
            }
            s.push(i);
        }
        return result;
    }
};